File: C++/Link List/Add Two Numbers II.cpp
/*
You are given two non-empty linked lists representing two non-negative integers. 
The most significant digit comes first and each of their nodes contain a single digit. 
Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Follow up:
What if you cannot modify the input lists? In other words, reversing the lists is not allowed.
*/


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
    private:
        int length_link_list(ListNode* input)
        {
            int length_link = 0;
            while(input)
            {
                length_link += 1;
                input = input->next;
            }
            return length_link;
        }
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int length_l1 = length_link_list(l1);
        int length_l2 = length_link_list(l2);
        int max_length = max(length_l1,length_l2);
        ListNode dummyNode(0);
        ListNode *prev_nine = NULL, *cur = &dummyNode;
        for (size_t i = 0; i < max_length; ++i){
            int val = 0;
            if ((max_length - i) <= length_l1)
            {
                val += l1->val;
                l1 = l1->next;
            }
            
            if ((max_length - i) <= length_l2)
            {
                val += l2->val;
                l2 = l2->next;
            }
            
            if (val > 9)
            {
                while (prev_nine != NULL && prev_nine != cur)
                {
                    prev_nine->val =  (prev_nine->val + 1) % 10;
                    prev_nine = prev_nine->next;
                }
                
                cur->val = (cur->val + 1) % 10;
                prev_nine = (cur->val == 9) ? cur : NULL;
            
            } 
            if (val % 10 == 9)
            {  
                prev_nine = prev_nine ? prev_nine : cur;
            }
             else
                prev_nine = NULL;
            cur->next = new ListNode(val % 10);
            cur = cur->next;
        }
        
        ListNode* result = NULL;
        if (dummyNode.val != 0){
            result = new ListNode(dummyNode.val);
            result->next = dummyNode.next;
        }
        else
            result = dummyNode.next;
        return result;
        
    }
};

File: Python/Link List/Add Two Numbers II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        def LenLink(self,root):
            Len = 0
            while root:
                root = root.next
                Len += 1
            return Len
        len1,len2 = LenLink(self,l1), LenLink(self,l2)
        maxlen = max(len1,len2)
        dummy = ListNode(0)
        prevnine = None
        current = dummy
        Result = None
        for i in range(0,maxlen):
            currentval = 0
            if maxlen-i<=len1:
                currentval += l1.val
                l1 = l1.next
            if maxlen-i <=len2:
                currentval += l2.val
                l2 = l2.next
            if currentval>9:
                while ( prevnine != None and prevnine != current):
                    prevnine.val = ( prevnine.val + 1 ) % 10
                    prevnine = prevnine.next
                current.val = (current.val + 1)% 10
                prevnine = current if currentval==9 else None
            
            if currentval %10 == 9:
                prevnine = prevnine if prevnine else current
            else:
                prevnine = None
            
            current.next = ListNode(currentval%10)  
            current = current.next
      
        if dummy.val != 0 :
            Result = ListNode(dummy.val)
            Result.next = dummy.next
        else :
            Result = dummy.next
        
        return Result
